fault/wrapper.py

```diff
@@ -49,11 +50,49 @@ class PortWrapper:
     def __init__(self, port, parent):
         self.port = port
         self.parent = parent
+        self.init_done = True
 
     def expect(self, value):
         select_path = self.select_path
         select_path.tester.expect(select_path, value)
 
+    def __setitem__(self, key, value):
+        if not isinstance(self.port, (m.ArrayType, m.TupleType)):
+            raise Exception(f"Can only use item assignment with arrays and "
+                            f"tuples not {type(self.port)}")
+        select_path = self.select_path
+        select_path[-1] = select_path[-1][key]
+        select_path.tester.poke(select_path, value)
+
+    def __getitem__(self, key):
+        if not isinstance(self.port, (m.ArrayType, m.TupleType)):
+            raise Exception(f"Can only use getitem with arrays and "
+                            f"tuples not {type(self.port)}")
+        select_path = self.select_path
+        return PortWrapper(self.port[key], self.parent)
+
+    def __getattr__(self, key):
+        try:
+            object.__getattribute__(self, "init_done")
+            if not isinstance(self.port, (m.TupleType)):
+                raise Exception(f"Can only use getattr with tuples, "
+                                f"not {type(self.port)}")
+            select_path = self.select_path
+            return PortWrapper(getattr(self.port, key), self.parent)
+        except AttributeError:
+            return object.__getattribute__(self, key)
+
+    def __setattr__(self, key, value):
+        try:
```

Comment:
I know it's clunky, but since this is internal and `__init__` is not that big, maybe just using `object.__setattr__()` in `__init__` and avoiding this try-catch here would work? Unless I'm missing something.

Reply:
We could also do the same in the other place where we need to do this hack.

Reply:
using setattr would work, but is that really better than this try catch? Both solutions suck, I don't think it's that much better. If we want to add fields to __init__ later, that solution is more annoying. This one works, a metaclass pipeline is probably the right way to do it, but I won't have time to investigate that until after Thursday.

Reply:
I think there were calls to setattr/getattr even before init (in the instance creation pipeline) which was another reason that overriding the default behavior even with using `getattr` in `__init__`, but I don't remember specifically (this is the result of hacking my way through to a working solution). You're welcome to explore an alternative.

Reply:
Fair enough, didn't realize the complexity behind that solution too. Fine with this for now.